functions/src/glean.ts

```diff
@@ -0,0 +1,228 @@
+ * studyUnenrollment
+ * Glean ping: study-unenrollment
+ */
+export async function studyUnenrollment(
+  rallyID: string,
+  studyID: string
+): Promise<void> {
+  const releaseGlean = await gleanLock.acquire();
+  initializeGlean();
+
+  rallyMetrics.id.set(rallyID);
+  unenrollmentMetrics.studyId.set(studyID);
+
+  await submitPingFlag.acquire();
+  rallyPings.studyUnenrollment.submit();
+  await submitPingFlag.waitForUnlock();
+
+  releaseGlean();
+}
+
+/*
+ * Helper function for initializing Glean
+ */
+function initializeGlean(): void {
+  Glean.setDebugViewTag(GLEAN_DEBUG_VIEW_TAG);
+  Glean.setLogPings(true);
+
+  // Glean.initialize is a no-op if Glean is already initialized
+  Glean.initialize(GLEAN_RALLY_APP_ID, true, {
+    appDisplayVersion: GLEAN_APP_DISPLAY_VERSION,
+    plugins: [new PingEncryptionPlugin(GLEAN_ENCRYPTION_JWK)],
+    httpClient: new CustomPingUploader(),
+  });
+}
+
+/*
+ * Helper function for setting user metrics
+ * from demographic data (mapping)
+ */
+function setUserMetrics(data: any): void { // eslint-disable-line @typescript-eslint/no-explicit-any
+  if ("age" in data) {
+    userMetrics.age[`band_${data["age"]}`].set(true);
+  }
+
+  if ("gender" in data) {
+    userMetrics.gender[data["gender"]].set(true);
+  }
+
+  if ("hispanicLatinxSpanishOrigin" in data) {
+    const label =
+      data["hispanicLatinxSpanishOrigin"] === "other"
+        ? "other"
+        : "hispanic_latinx_spanish";
+    userMetrics.origin[label].set(true);
+  }
+
+  if ("race" in data) {
+    for (const raceLabel of data["race"]) {
+      const label =
+        raceLabel === "american_indian_or_alaska_native"
+          ? "am_indian_or_alaska_native"
+          : raceLabel;
+      userMetrics.races[label].set(true);
+    }
+  }
+
+  if ("school" in data) {
+    const KEY_FIXUP: any = { // eslint-disable-line @typescript-eslint/no-explicit-any
+      high_school_graduate_or_equivalent: "high_school_grad_or_eq",
+      some_college_but_no_degree_or_in_progress: "college_degree_in_progress",
+    };
+
+    const originalLabel = data["school"];
+    const label =
+      originalLabel in KEY_FIXUP ? KEY_FIXUP[originalLabel] : originalLabel;
+    userMetrics.school[label].set(true);
+  }
+
+  if ("exactIncome" in data) {
+    userMetrics.exactIncome.set(data["exactIncome"]);
+  }
+
+  if ("zipcode" in data) {
+    userMetrics.zipcode.set(data["zipcode"]);
+  }
+}
+
+/**
+ * Custom Ping Uploader for Glean
+ * TODO: replace direct POST request with Google Cloud Task
+ */
+class CustomPingUploader extends Uploader {
+  async post(
+    url: string,
+    body: string | Uint8Array,
+    headers: Record<string, string>
+  ): Promise<UploadResult> {
+    const result = await axios
+      .post(url, body, { headers: headers })
+      .then(function (response) {
+        return {
+          status: response.status,
+          result: UploadResultStatus.Success,
+        };
+      })
+      .catch(function (error) {
+        console.log(error);
```

Comment:
I don't think a `console.log` will show up in our monitoring, you should consider using [the Google cloud functions logger](https://firebase.google.com/docs/functions/writing-and-viewing-logs).